### ISSUE TITLE
Feature suggestion: visually distinguish color cells with a tag

### ISSUE THREAD

**Issue by SultanOrazbayev** [jupyterlab/jupyterlab#8457]
Right now, to infer whether a cell has a tag one needs to open the sidebar and go over the cells that have associated tags. It would be great if the cells that have a tag are visually distinct (e.g. different cell background, different cell border size/color).

If there is a way to achieve this already, then kindly let me know.

**Comment by telamonian** [comment]
I like this idea. +1 to the concept of being able to set a color on a particular tag, and then somehow marking tagged cells with that same color

**Comment by ricwillis98** [comment]
I'd like to work on this

Should the color be specified with a text input? A drop down?

I think a different cell border would be good for visually distinguishing them.

**Comment by jtpio** [comment]
For reference, a similar idea was recently shared on Discourse: https://discourse.jupyter.org/t/styling-cells-through-metadata/4978

With classic notebook extensions:

- https://github.com/innovationOUtside/nb_cell_execution_status
- https://github.com/innovationOUtside/nb_extension_tagstyler

**Comment by ricwillis98** [comment]
What should happen when multiple tags with associated colors are added to the cell? Just pick one for now?

**Comment by SultanOrazbayev** [comment]
Thanks, @ricwillis98, this is looking promising! 

For multiple colors, I like telamonian's suggestion in your PR (having multiple 'stripes'/mini-badges). Your initial suggestion also looks good to me, in case of multiple tags, it's possible to introduce stripes to the left side as in the attached image... but any visual distinction would be great.

**Comment by echarles** [comment]
Would be great to reuse the jupyterlab color palette defined in the CSS for more homogenous look and fell.

**Comment by ricwillis98** [comment]
@echarles good suggestion, it's a bit of a work around but you can by specifying the color as you would in css e.g. tag1;var(--jp-brand-color0)

This is how it looks in my current version

![image](https://user-images.githubusercontent.com/62662248/86282734-6b27f480-bba5-11ea-809a-1068bb556a8e.png)

**Comment by SultanOrazbayev** [comment]
The original use case for me is the "parameters" tag used by `papermill`. It's not always obvious whether a notebook has it, and so it's great to be able to see it right away. However, this adds a constraint of what the value of the tag can be (`parameters`). I understand that this will be implemented in a later version, but thought I'd mention it just in case.

Thanks for your work, @ricwillis98!

**Comment by echarles** [comment]
@ricwillis98 Great workaround!

**Comment by jasongrout** [comment]
The capabilities here look really valuable. Thanks for exploring this!

I'm concerned that this approach of encoding the colors in the tag names:

1. It is mixing concerns - it's mashing together semantic content (the tag name) with its presentation (the color). What if I use colors differently than you, or need to change one of the colors (for example, I use a different theme)?
2. If we merge this into core JupyterLab, it is also setting a de facto nbformat standard, without a broader discussion in the nbformat community.

What if we instead made a plugin in JupyterLab that exposed a tag:color mapping setting (or even made it part of the jlab notebook plugin)? It would then become a convenience in JupyterLab to display tags, rather than being stored in the notebook data (thus not needing a discussion about nbformat standards), and makes it easy to customize the colors in your JupyterLab settings (thus not mixing the semantic and presentation information).

**Comment by ricwillis98** [comment]
@jasongrout I agree, I'm not attached to the syntax I'm using right now, it was a temporary fix while I work on actually displaying the tag information and then I/we can come up with some better UI/method to change the colors. 

I'm not sure what you mean by 2. (new contributor to Jupyter). Is it altering the notebook appearance by showing tag colors in the notebook that is the problem or something else? 

As this idea has evolved, I was actually starting to wonder if it's better (for now) to assign random jupyter UI/brand colors as suggested by @echarles to every tag without the option to specify them manually. The colors would be stored in cell.model.metadata along with tags. This way we could add the feature of visually distinguishing cells by tag and then address the UI for choosing the color later. Does this address your concern in 2.? 

What's the best way forward? How should we start the discussion you suggest in 2. or should I work on a plugin instead?

**Comment by SultanOrazbayev** [comment]
> As this idea has evolved, I was actually starting to wonder if it's better (for now) to assign random jupyter UI/brand colors as suggested by @echarles to every tag without the option to specify them manually.

This sounds like a good solution for now.

**Comment by jasongrout** [comment]
Thanks for the clarification.

> I'm not sure what you mean by 2. (new contributor to Jupyter). Is it altering the notebook appearance by showing tag colors in the notebook that is the problem or something else?

My concern here is storing the colors in the notebook file (i.e., cell metadata or notebook metadata) will start to standardize how such colors are to be stored in the notebook file, without discussion with other projects that deal with notebook files. I think we can avoid that discussion now by not storing the colors in the notebook file (i.e., keep the tag:color mapping as a JupyterLab setting, to be applied across all of JupyterLab, as a JLab specific convenience), or I think we can have that discussion on the nbformat repo with others that deal with notebook files to move towards standardizing how colors associated with tags are stored in a notebook file.

**Comment by jasongrout** [comment]
As for where the mapping is stored, do you expect a tag:color mapping will remain consistent for one user, or consistent in one notebook no matter who is using it? That determines whether it should be stored in my environment (i.e., as a jlab setting), or be stored in the file (i.e., as notebook metadata that is consistent no matter who is viewing the notebook).

My guess is that it is more natural to have the tag:color mapping be consistent for a user (so stored as a jlab setting).

**Comment by ricwillis98** [comment]
@jasongrout I was actually thinking it should be consistent for all users of the same notebook. I wouldn't want to open the notebook and a tag be blue for me and red for someone else etc. 
Continuing the idea of assigning a random color to a tag, what if a color was assigned to all tags by hashing the tag name some how (e.g. sum ascii value of each character and then modulo the number of color choices)? That way the colors would be consistent for everybody _and_ we wouldn't have to deal with the syntax/method for assigning a color just yet. What do you think?

**Comment by telamonian** [comment]
> Continuing the idea of assigning a random color to a tag, what if a color was assigned to all tags by hashing the tag name some how (e.g. sum ascii value of each character and then modulo the number of color choices)?

I love this idea, it's a clever idea, it's a cute idea. It may be too cute though for its own good. Let's try it out. The other obvious alternative is to just assign a color to each tag based on its index in the list of tags of the open notebook. This would satisfy the condition of having the same colors for all users of a particular notebook, though the same tag in different notebooks may have different colors.

**Comment by telamonian** [comment]
I've been thinking about the color issues here, as well as your comments @jasongrout, for a couple of days now. Here's my conclusions.

- For the  tag colors let's have each theme define a color cycle that can be used with the tags (and possibly elsewhere too), a la [the color cycles in matplotlib](https://matplotlib.org/3.2.2/users/dflt_style_changes.html#colors-color-cycles-and-color-maps).
  - Optionally, each theme may explicitly specify a set of `--jp-color-cycle0`, `--jp-color-cycle1`, etc vars
  - Otherwise, we'll harvest a color cycle from some set of the already existing theme vars

- For the user settings
  - Let's have one `"colored-tags"` option in the "Notebook" settings that controls whether the colored tags feature as a whole is on/off.
  - For the tag color choices, since the colors are ultimately coming from the themes, let's add a `"tag-colors"` option to the rest of the theme overrides in the "Theme" settings

**Comment by psychemedia** [comment]
Did this ever get as far as being integrated or packaged as an extension anywhere?

**Comment by vidartf** [comment]
See also #8298.

**Comment by psychemedia** [comment]
I have an extension here — https://github.com/innovationOUtside/jupyterlab_empinken_extension — which defines a few custom tags and associates specific styling with them.

**Comment by psychemedia** [comment]
Years on, and I note my color-cells-by-tag powered [extension](https://github.com/innovationOUtside/jupyterlab_empinken_extension) has rotted as far as JupyterLab v4 goes. Is there any likelihood of JupyterLab v4 propagating a cell tag through to HTML? 

I note that notebook v7 finally relented to allowing [custom css](https://github.com/jupyter/notebook/pull/6841) files to be trivially added. Is this also possible in JupyterLab, which would simplify styling if tags were in HTML classes? I have previously used a hacky [minimal extension](https://blog.ouseful.info/2022/03/28/finally-simple-custom-styled-markdown-cells-in-jupyterlab/) to try to add custom style files.

**Comment by jtpio** [comment]
> Is this also possible in JupyterLab, which would simplify styling if tags were in HTML classes?

Not yet I believe, but the work done in https://github.com/jupyter/notebook/pull/6841 could actually be moved to JupyterLab.

This would fix https://github.com/jupyterlab/jupyterlab/issues/12318.

It was implemented in Notebook 7 first because this was supported in the Classic Notebook. But if there is interest yes it would be great to add support for it in JupyterLab too (and remove the code in Notebook 7). Maybe this could even be added to JupyterLab 4.1.

cc @RRosio who might be interested
